homeassistant/components/emulated_hue.py

```diff
@@ -226,58 +241,49 @@ def __init__(self, hass, config):
         self.config = config
         self.cached_states = {}
 
+    @core.callback
     def get(self, request, username, entity_id=None):
         """Handle a GET request."""
         if entity_id is None:
-            return self.get_lights_list()
+            return self.async_get_lights_list()
 
-        if not request.base_url.endswith('state'):
-            return self.get_light_state(entity_id)
+        if not request.path.endswith('state'):
+            return self.async_get_light_state(entity_id)
 
-        return self.Response("Method not allowed", status=405)
+        return web.Response(text="Method not allowed", status=405)
 
+    @asyncio.coroutine
     def put(self, request, username, entity_id=None):
         """Handle a PUT request."""
-        if not request.base_url.endswith('state'):
-            return self.Response("Method not allowed", status=405)
-
-        content_type = request.environ.get('CONTENT_TYPE', '')
-        if content_type == 'application/x-www-form-urlencoded':
-            # Alexa sends JSON data with a form data content type, for
-            # whatever reason, and Werkzeug parses form data automatically,
-            # so we need to do some gymnastics to get the data we need
-            json_data = None
-
-            for key in request.form:
-                try:
-                    json_data = json.loads(key)
-                    break
-                except ValueError:
-                    # Try the next key?
-                    pass
-
-            if json_data is None:
-                return self.Response("Bad request", status=400)
-        else:
-            json_data = request.json
+        if not request.path.endswith('state'):
+            return web.Response(text="Method not allowed", status=405)
+
+        if entity_id and self.hass.states.get(entity_id) is None:
+            return self.json_message('Entity not found', HTTP_NOT_FOUND)
+
+        try:
+            json_data = yield from request.json()
+        except ValueError:
+            return self.json_message('Invalid JSON', HTTP_BAD_REQUEST)
 
-        return self.put_light_state(json_data, entity_id)
+        result = yield from self.async_put_light_state(json_data, entity_id)
+        return result
 
-    def get_lights_list(self):
+    def async_get_lights_list(self):
         """Process a request to get the list of available lights."""
         json_response = {}
 
-        for entity in self.hass.states.all():
+        for entity in self.hass.states.async_all():
             if self.is_entity_exposed(entity):
                 json_response[entity.entity_id] = entity_to_json(entity)
 
         return self.json(json_response)
 
-    def get_light_state(self, entity_id):
+    def async_get_light_state(self, entity_id):
```

Comment:
This one and the above one are both helper methods for the real view handlers. I've labelled them both callback. Neither will ever be called by other code than the coroutine handlers in this view.